### ISSUE TITLE
resolve relative URLs in e-* html

### ISSUE THREAD

**Issue by willnorris** [microformats/microformats2-parsing#38]
The parsing spec does not currently include any special handling of URLs in the `html` value for e-* microformats.  From <http://microformats.org/wiki/microformats2-parsing#parsing_an_e-_property>:

> `html`: the `innerHTML` of the element by using the HTML spec: Serializing HTML Fragments algorithm, with leading/trailing whitespace removed.

However, some of the microformats tests do resolve relative URLs.  See for example:
 - https://github.com/microformats/tests/blob/master/tests/microformats-v2/h-entry/urlincontent.json
 - https://github.com/microformats/tests/blob/master/tests/microformats-mixed/h-entry/mixedroots.json

The major libraries are somewhat split on this.  [PHP](https://php.microformats.io/?url=https%3A%2F%2Frawgit.com%2Fmicroformats%2Ftests%2Fmaster%2Ftests%2Fmicroformats-v2%2Fh-entry%2Furlincontent.html) and [Ruby](https://ruby.microformats.io/microformats?utf8=%E2%9C%93&url=https%3A%2F%2Frawgit.com%2Fmicroformats%2Ftests%2Fmaster%2Ftests%2Fmicroformats-v2%2Fh-entry%2Furlincontent.html) **do** resolve relative URLs.  [Go](https://go.microformats.io/?url=https%3A%2F%2Frawgit.com%2Fmicroformats%2Ftests%2Fmaster%2Ftests%2Fmicroformats-v2%2Fh-entry%2Furlincontent.html), [Python](https://python.microformats.io/?url=https%3A%2F%2Frawgit.com%2Fmicroformats%2Ftests%2Fmaster%2Ftests%2Fmicroformats-v2%2Fh-entry%2Furlincontent.html&parser=html5lib), and [Node](https://glennjones.net/tools/microformats/?url=https%3A%2F%2Frawgit.com%2Fmicroformats%2Ftests%2Fmaster%2Ftests%2Fmicroformats-v2%2Fh-entry%2Furlincontent.html&callback=&dateformat=auto) **do not** resolve relative URLs.

[Recent discussion in #microformats](https://chat.indieweb.org/microformats/2018-08-17#t1534534160369100) was inconclusive (though we didn't explore it too deeply)

At the very least, we need to synchronize the spec and the test cases.  Personally, I'm leaning toward updating the spec to resolve relative URLs, since otherwise they are useless in any kind of embedding use-case, and they may not actually be able to be resolved, since you no longer have the `<base>` element.

**Comment by aaronpk** [comment]
The lack of the `<base>` element in the parsed result, as well as the fact that some parsers already do this, make me lean towards adding it to the spec as well.

**Comment by Zegnat** [comment]
I agree. My first reaction was to keep the HTML as close to the original as possible (including relative URLs), but the embedding use-case won me over! The relative URLs cannot stand on their own in the HTML fragment, as we would have lost meaning. Resolving them makes a lot of sense!

**Comment by Zegnat** [comment]
Because these issues tend to grow stale if there is no concrete proposal, here is one.

I propose replacing this line in [the parsing specification for `e-*`](http://microformats.org/wiki/microformats2-parsing#parsing_an_e-_property):

> * `html`: the `innerHTML` of the element by using the [HTML spec: Serializing HTML Fragments algorithm](https://html.spec.whatwg.org/multipage/parsing.html#serialising-html-fragments), with leading/trailing whitespace removed. 

With:

> * `html`: the results of running the [HTML fragment serialization algorithm](https://html.spec.whatwg.org/multipage/parsing.html#serialising-html-fragments) from the HTML specification on the element, with:
>    
>    * relative URLs in [HTML attributes that take URLs as values](https://html.spec.whatwg.org/multipage/indices.html#attributes-3) replaced with matching normalized absolute URLs, following the containing document’s language’s rules for resolving relative URLs;
>    * leading/trailing whitespace removed;

The attributes table in the HTML specification is one of the few resources that tells us where URLs are, without having to leave the parser implementer guessing about what strings in the HTML fragment constitute relative URLs. This came out of a discussion with regards to the Webmention spec (w3c/webmention#91).

The “… following the containing document’s language’s rules …” phrasing is taken straight from other places where we resolve URLs.

I have kept the “with” phrasing because I am not sure how to best sum up the actions that need to be taken by the parser. Most implementers will probably first resolve the URLs, then run the serialisation algorithm, and then remove whitespace. That’s a slightly different order from what this proposal suggests. Let me know if there are better ways to phrase this.

**Comment by sknebel** [comment]
Looks good.

Can anyone think of use-cases that would prefer URLs not to be normalized? Theoretical example I can come up with: recovery of content from a site with mf2 markup, where relative URLs to pictures, other posts, ... might be preferred. 

If that's an issue, the parser should at least return the found base URL for the page, so later steps can resolve if they want. In the majority of cases, the HTML from the `e-*` properties has to be postprocessed anyways (filtering safe tags, replacing images with proxied versions, ...), and resolving would then add "just" another step to that.

**Comment by gRegorLove** [comment]
Sounds good. I'm +1 on @Zegnat's proposal.

**Comment by willnorris** [comment]
[@Zegnat's proposal](https://github.com/microformats/microformats2-parsing/issues/38#issuecomment-414262178) has been [implemented in the go parser](https://github.com/willnorris/microformats/blob/master/microformats.go#L119-L140), though it is not currently enabled.

**Comment by snarfed** [comment]
This came up recently in https://github.com/snarfed/bridgy-fed/issues/390#issuecomment-1408879534. php-mf2 currently resolves relative URLs in `e-*`, mf2py doesn't. Shall we try again to get @Zegnat's proposal ^ into the spec? Also, mf2py and other parser contributors, any thoughts? @sknebel @tommorris @angelogladding

**Comment by jgarber623** [comment]
[MicroMicro](https://github.com/jgarber623/micromicro) resolves relative URLs within `e-*` properties. I don’t remember building it that way for any other reason than to pass the microformats2 test suite.

**Comment by sknebel** [comment]
mf2py has this ready to go (and thus I guess votes in favor of this change)

**Comment by jalcine** [comment]
This is supported by the Rust Microformats parser and [is demonstrated in the tested documentation](https://docs.rs/microformats/latest/microformats/fn.resolve_text_from_html.html), as this is done generically with plain text.

(Originally published at: https://jacky.wtf/2023/11/u1y4)

**Comment by jalcine** [comment]
I'm also in favor of this change!

(Originally published at: https://jacky.wtf/2023/11/JeNH)
